elm-street.cabal

```diff
@@ -130,7 +130,7 @@ test-suite elm-street-test
                      , types
                      , aeson
                      , bytestring >= 0.10 && <= 0.12
-                     , hspec ^>= 2.7.1
+                     , hspec >= 2.7.1 && <= 2.10
```

Comment:
Based on some recent discussions, at the very least for the test suite, shouldn't we just drop the upper bound constraint?

Reply:
I agree. Done.

Reply:
Afaik even stackage is running tests for libraries.

I think upper bound could be dropped but we would also need to have stronger guarantee with backward compatibility. Because if we release library with relaxed upper bound and it breaks the build using cabal resolver the only real fix is to release never version which is backward compatible with all the previously supported libraries as well as new ones. Something like this might need more discussion than just our recent one. If we would just relax upper bound we can occasionally ran into situation where resolver would resolve to something that won't compile and since hackage is immutable the only way to resolve is to release version that is compatible with new thing on top of supporting what previous version did support.

Reply:
That should not be an issue, because hackage has the feature of candidate releases (which our hackage cookbook in notion encourages us to use), where you can make sure the build (including haddocks build) passes before you "publish" the release. So I think this shouldn't be an issue.

Reply:
And I actually tested with latest versions of packages available, so I know there are no (as of today) newer versions of dependencies that we removed upper bounds from, that would break this package.

Reply:
> That should not be an issue, because hackage has the feature of candidate releases (which our hackage cookbook in notion encourages us to use),

I know - I wrote that original version for hackage cookbook. Anyway package candidate doesn't solve potential breakages in future versions stackage. When package fails to build either due to build problem or constrains problem it's blacklisted away from stackage. For stackage this is not big deal as it's based on package set but cabal resolver might be picking version of package which can't build for ever - unless you publish version that will be favorable for resolver (restricting upper bound retrospectively is not an option because cabal would pick the previous relaxed version). That's the danger of loosening upper bound.

Reply:
As far as I know elm-street is not on stackage. Are we even trying to get it there?
I understand the dangers involved in removing the version bounds.
But from what you said before it seemed to me that you are *against* setting upper bound on version because of the annoyances with constantly having to bump the bounds as new versions of dependencies are released.
So what camp are you in? :smile:

Reply:
yes! You're right I'm for relaxing upper bounds. Just wanted to point out that it comes with responsibility - if we do this we need to be aware we will have to put in some effort in case we break downstream builds in the future and resolve these accordingly.

BTW I wonder why we don't publish our packages to stackage given how much we rely on it internally.